Operate First - Minimizing RBAC Requirements

eletonia opened the pull request:
This PR adds namespaced roles in the Helm templates instead of a cluster role to minimize RBAC requirements to deploy M4D on Operate First

eletonia commented:
@roee88 I changed the `m4d-blueprints` role to have access to all resources, added role bindings for the `m4d-system` and `m4d-blueprints` roles, and added cluster roles for `m4dapplication`, `batchtransfer`, and `streamtransfer` Can you review these changes?

eletonia commented:
The `m4d-blueprints` and `m4d-system` roles and role bindings now check for `coordinator` and the m4dapplication cluster role now checks for `worker` @roee88 Does this look correct?

roee88 commented on charts/m4d/templates/m4dapplication-cr.yaml:
```diff
@@ -0,0 +1,16 @@
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: m4dapplication-editor
```
Consider renaming to `m4d-user` or something similar to reflect that it's a role for users of M4D

eletonia commented on charts/m4d/templates/m4d-system-role.yaml:
```diff
@@ -0,0 +1,25 @@
+  - plotters
+  - m4dstorageaccounts
+  - m4dmodules
+  - datasets
```
I checked `manager-cr.yaml` and only plotters has a `/status` so I just added that to this file

eletonia commented:
Thanks for reviewing @roee88 
I made the suggested changes and added the new `m4d-admin` cluster role
Can we merge?

roee88 commented:
> Thanks for reviewing @roee88
> I made the suggested changes and added the new `m4d-admin` cluster role
> Can we merge?

Approved. I'll check it manually next week since the CI is still broken and then merge.

ronenkat commented:
@eletonia can you rebase? let see if we pass all tests

roee88 commented on charts/m4d/templates/m4d-blueprints-rb.yaml:
```diff
@@ -0,0 +1,16 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: {{ template "m4d.fullname" . }}-blueprints-rb
```
Missing namespace, should be set to `m4d-blueprints`. Thanks @shlomitk1.

eletonia commented:
@ronenkat @roee88 I rebased and added the `m4d-blueprints` namespace to the m4d-blueprints role binding.
Now the integration tests are failing for batch and stream transfers. The `streamtransfer-sample` secret and `batchtransfer-sample` job are not being created for some reason, do you know why?

shlomitk1 commented:
> 
> 
> @ronenkat @roee88 I rebased and added the `m4d-blueprints` namespace to the m4d-blueprints role binding.
> Now the integration tests are failing for batch and stream transfers. The `streamtransfer-sample` secret and `batchtransfer-sample` job are not being created for some reason, do you know why?

The yamls in manager/testdata (batchtransfer and streamtransfer) should be invoked in m4d-blueprints rather than in default namespace.

eletonia commented:
> > @ronenkat @roee88 I rebased and added the `m4d-blueprints` namespace to the m4d-blueprints role binding.
> > Now the integration tests are failing for batch and stream transfers. The `streamtransfer-sample` secret and `batchtransfer-sample` job are not being created for some reason, do you know why?
> 
> The yamls in manager/testdata (batchtransfer and streamtransfer) should be invoked in m4d-blueprints rather than in default namespace.

I changed the namespace in [batchtransfer.yaml](https://github.com/eletonia/the-mesh-for-data/blob/operate-first-rbac/manager/testdata/batchtransfer.yaml) and [streamtransfer.yaml](https://github.com/eletonia/the-mesh-for-data/blob/operate-first-rbac/manager/testdata/streamtransfer.yaml) to `m4d-blueprints` but now the unit tests are failing since it can't find the `m4d-blueprints` namespace:
```
Should simulate a StreamTranssfer successfully [It]
    /home/runner/work/mesh-for-data/mesh-for-data/manager/controllers/motion/streamtransfer_controller_test.go:47

    Expected success, but got an error:
        <*errors.StatusError | 0xc0003e2a00>: {
            ErrStatus: {
                TypeMeta: {Kind: "", APIVersion: ""},
                ListMeta: {
                    SelfLink: "",
                    ResourceVersion: "",
                    Continue: "",
                    RemainingItemCount: nil,
                },
                Status: "Failure",
                Message: "namespaces \"m4d-blueprints\" not found",
                Reason: "NotFound"
```

roee88 commented:
@shlomitk1 take a look at https://github.com/kubernetes-sigs/controller-runtime/pull/1435.

shlomitk1 commented:
> 
> 
> @shlomitk1 take a look at [kubernetes-sigs/controller-runtime#1435](https://github.com/kubernetes-sigs/controller-runtime/pull/1435).

It is supported starting from v0.9.0. This breaks dependencies, I've tried to upgrade other k8s dependencies but it causes problems with required dependencies for datashim...
